### PR TITLE
Final project submission

### LINKED ISSUE #8: [3pt] Recipe details #Member 2
- Should display a toggle button that allows for a recipe to be made public or private.
- If the recipe is public or the user is the owner of the recipe, should display the recipe details as in the wireframe.
- If the user is the owner of the recipe, should lead to the form that allows the user to add new food.

### LINKED ISSUE #9: [1pt] Fixing n+1 problem # Group task
- Double check for n+1 problems

### PR DESCRIPTION
**Pull request by Marcraphael12**
In this project, we have to build a fully functional recipe app with Ruby and RoR following the requirements bellow.
### Project requirements

- You should follow the layout of the wireframes provided. You should personalize the rest of the design including colors, typographies, spacings, etc.
- Login page and registration page:
    - Should be built with Devise.
- Recipes list:
    - Should display a list of recipes created by the logged-in user as in the wireframe.
    - Should lead to recipe details.
- Public recipe list:
    - Should display a list of all public recipes ordered by newest as in the wireframe.
    - Should lead to recipe details.
    - If the user is the owner of the recipe, should allow the user to delete it.
- Recipe details:
    - Should display a toggle button that allows for a recipe to be made public or private.
    - If the recipe is public or the user is the owner of the recipe, should display the recipe details as in the wireframe.
    - If the user is the owner of the recipe, should lead to the form that allows the user to add new food.

- Make sure there are no N+1 queries happening.
- Create a navigation menu that allows users to open all of the pages you created.
- Write unit and integration tests

- Food list:
    - Should display a list of food added by the logged-in user as in the wireframe (display also quantity of a given food).
    - Should lead to a form that allows users to add new food.
- General shopping list view:
    - Should show the list of food that is missing for all recipes of the logged-in user (compare the list of food for all recipes with the general food list of that user).
    - Should count the total food items and total price of the missing food.

**Technical setup**
- Set up the repository on GitHub and use Gitflow.
- Set up Devise for authentication.
- Set up RSpec and Capybara testing libraries.

### Workload distribution

To tackle this challenge, you need to create a Kanban board with a GitHub project that translates the requirements into a set of tasks that you will be able to use to organize your work. You will create your board in a separate activity.

You will be working in this way:
- All tasks should have time estimates and your job is to distribute them between team members in a fair way.
- The common tasks will be divided among all of you or completed as a team (pair programming).
- All tasks should be based on the cards from your Kanban board.

## Work and submission mode

- You should implement the above requirements only in **one repository** in your group.
- You should ask for a review and submit this activity **on behalf of your group.**

## Code review

You will give and receive code reviews from your teammates. Each task should have a separate pull request that is reviewed by one of your teammates.
Once the entire project is ready, one of your team members will request a code review on behalf of your group.

### CODE REVIEW

File: app/views/recipes/show.html.erb [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    <div>
+      <label for="checkbox">Public</label>
+      <label class="switch">
+        <input type="checkbox" class="toggle" name="checkbox" <%= @recipe.public ? 'checked' : '' %> />
+        <span class="slider round"></span>
+      </label>
+    </div>
+  </div>
+  <div class="link-box">
+    <%= link_to 'Generate shopping list' %>
```
Comment (Qoosim) [inline app/views/recipes/show.html.erb]: - [ ]  It is required that on the recipe show page, you should be able to generate a shopping list that matches the given [wireframe](https://raw.githubusercontent.com/microverseinc/curriculum-rails/main/recipe-app/images/general_shopping_list.png?token=GHSAT0AAAAAACJ3BMNOEQO5DQQQQZ56Z35AZKZF2XA). :point_down: with this url `/general_shopping_list`

shopping-list

Nothing seems to happen when I clicked on that link. Kindly look into it and fix the issue.

[shopping-list_remove.webm](https://github.com/Gh-Reza/RecipeApp/assets/26597432/129b725e-0f9e-43c7-a900-2f819d683be8)

File: app/views/recipes/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      <tbody>
+        <% @recipe_foods.each do |item| %>
+          <tr>
+            <td><%= item.food.name %></td>
+            <td><%= item.quantity %> <%= item.food.measurement_unit %></td>
+            <td><%= item.food.price * item.quantity %></td>
+            <td>
+              <%= button_to "Edit"%>
+              <%#= link_to 'Delete', recipe_food_path(recipe_id: @recipe.id, id: item.food.id), method: :delete, data: { confirm: 'Are you sure?' } %>
+              <%= button_to "Remove", recipe_food_path(item.id), method: :delete %>
```
Comment (Qoosim) [inline app/views/recipes/show.html.erb]: - [ ]  I should be able to remove a recipe item if I am the owner. This too seems not to be working. Kindly look into it and fix where the issue is. Watch the video above to see this in action. :point_up_2:

File: app/views/recipes/public_recipes.html.erb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+          <p><%= recipe.user.name %></p>
+        </div>
+        <div>
+          <p>Total items: <%= recipe.foods.count %></p>
+          <p>Total price: <%= recipe.total_foods_price %></p>
+        </div>
+      </li>
+    <% end %>
+  </ul>
+</section>
```
Comment (Qoosim) [inline app/views/recipes/public_recipes.html.erb]: -  [  ] It is required that your public recipe list should look like the given wireframe. You design looks a bit different from the given wireframe. Kindly look into this and fix it.

### Your Design

your-public-recipe

### Given Wireframe

public-recipe

File: spec/views/food_spec.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+      visit foods_path
+    end
+
+    it 'can see all the food' do
+      expect(page).to have_content 'Test food 1'
+      expect(page).to have_content 'Test food 2'
+      expect(page).to have_content 'Test food 3'
+      expect(page).to have_content 'Test food 4'
+      expect(page).to have_content 'Test food 5'
+    end
```
Comment (Qoosim) [inline spec/views/food_spec.rb]: - [ ]  Kindly write dynamic tests for your app. These were hardcoded.

File: spec/views/public_recipes_spec.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    end
+
+    it 'can see all the public recipes' do
+      expect(page).to have_content 'Test recipe 1'
+      expect(page).to have_content 'Test recipe 2'
+      expect(page).to have_content 'Test recipe 3'
+      expect(page).to have_content 'Test recipe 4'
+      expect(page).to have_content 'Test recipe 5'
+    end
+
```
Comment (Qoosim) [inline spec/views/public_recipes_spec.rb]: - [ ]  Kindly write dynamic tests for your app. These were hardcoded.

File: spec/views/recipe_spec.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+      visit recipes_path
+    end
+
+    it 'can see all the recipes' do
+      expect(page).to have_content 'Test recipe 1'
+      expect(page).to have_content 'Test recipe 2'
+      expect(page).to have_content 'Test recipe 3'
+      expect(page).to have_content 'Test recipe 4'
+      expect(page).to have_content 'Test recipe 5'
+    end
```
Comment (Qoosim) [inline spec/views/recipe_spec.rb]: - [ ]  Kindly write dynamic tests for your app. These were hardcoded.

**Review by Qoosim — CHANGES_REQUESTED** [review]
#  *Status: Required Changes* :recycle:

Hi @team, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
-  Implemented some project requirements. :heavy_check_mark: 
-  No linter errors. :heavy_check_mark: 
-  Good README file. :heavy_check_mark: 
-  Correct Gitflow. :heavy_check_mark: 




## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by ridwanediallo — CHANGES_REQUESTED** [review]
# Changes Required ♻️
## Hi team, 


AlmostThereReachGIF

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


## (Highlights) Good Points: :+1:
- You documented your work professionally. ✔
- GitHub flow is used correctly. ✔
- All linters checks are ok. ✔


 ## Required Changes ♻️

- [ ] So far, you've done a good job of attempting to implement the changes requested by the previous reviewer. However, when I attempt to run your project, I receive the error shown in the screenshot below :point_down: This indicates that there is no route helper method named public_recipes_path in your Rails application. This usually occurs when no corresponding route is defined in your config/routes.rb file. 
To resolve this issue, you need to ensure that you have a route defined for the "Public recipes" resource in your config/routes.rb file. Here's an example of how you can define a route for a resource: `resources :public_recipes`.
### Issue :point_down: 

Screenshot from 2023-11-19 00-25-05

- [ ] After resolving the issue and seeing if I could continue with the review, I encountered another error when attempting to create a recipe. As a result, I was unable to determine whether you had completed the changes requested by the previous reviewer. Please fix those issues and ensure that you addressed all of the changes requested by the previous reviewer.  :ok_hand: 

### Error creating a recipe :point_down: 

Screenshot from 2023-11-19 00-24-22

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by prolajumokeoni — APPROVED** [review]
Hi @Team

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
## Highlights
  - App works properly✅ 
 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification @prolajumokeoni.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**